torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -693,15 +702,19 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
 
       auto bucket_view =
           replica.contents.narrow(0, offset, length).view(variable.sizes());
-      auto& grad = variable.grad();
-
-      // If a parameter is globally unused, we keep its grad untouched.
-      if (!global_unused) {
-        if (!grad.defined()) {
-          grad = at::empty(bucket_view.sizes(), bucket_view.options());
+      runGradCallbackForVariable(variable, [&](auto& grad) {
+        // If a parameter is globally unused, we keep its grad untouched.
+        if (!global_unused) {
+          if (!grad.defined()) {
+            grad = at::empty(bucket_view.sizes(), bucket_view.options());
+          }
+          grad.copy_(bucket_view);
+          // The grad is modified and needs to be written back.
+          return true;
         }
-        grad.copy_(bucket_view);
-      }
+        // The grad is not modified.
+        return false;
+      });
```

Comment:
Just realized there might be a problem here. This piece of code does not run in the post hook and hence the thread local will not be set here. We need to set the thread local in this callback as well: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L395. I'm wondering why this wasn't caught by the unit tests? Or maybe I'm missing something here?

Reply:
The thread local is already retrieved at this point. It's already stored in `Reducer::RpcContext`. The retrival happens when the post hook is called.